File: src/mahjong/models/meld.py
from dataclasses import dataclass
from typing import Tuple

from mahjong.constants.enums import ActionType
from mahjong.models.tile import Tile


@dataclass(frozen=True)
class Meld:
    """
    表示明牌组合（碰、杠）

    使用 frozen=True 确保不可变性。
    tiles 使用 Tuple 而非 List，保持类型一致性：
    - Tuple 是不可变的，与 frozen=True 语义一致
    - List 是可变的，会破坏不可变保证

    Attributes:
        meld_type: 明牌类型（PONG, KONG_EXPOSED, KONG_CONCEALED, KONG_UPGRADE）
        tiles: 牌的元组（不可变）
        is_concealed: 是否暗杠（用于判断门清）
            - True: 暗杠（不破坏门清）
            - False: 碰/明杠/补杠（破坏门清）
    """
    meld_type: ActionType
    tiles: Tuple[Tile, ...]
    is_concealed: bool = False

File: frontend/src/types/meld.ts
/**
 * Meld (exposed tile combination) types
 *
 * Backend correspondence: src/mahjong/models/meld.py::Meld
 */

import { createTile } from './tile';
import type { Tile } from './tile';

/**
 * Meld type enum
 *
 * Backend correspondence: ActionType enum subset
 * Note: Backend uses ActionType for all action types, frontend extracts meld-related types
 */
export enum MeldType {
  PONG = 'PONG',                   // 碰 (3 tiles)
  KONG = 'KONG',                   // 通用杠 (backward compatible)
  KONG_EXPOSED = 'KONG_EXPOSED',   // 明杠/直杠 (3 hand + 1 discarded, 2x score)
  KONG_CONCEALED = 'KONG_CONCEALED', // 暗杠 (4 hand, 1x score)
  KONG_UPGRADE = 'KONG_UPGRADE'    // 补杠/巴杠 (pong upgraded to kong, 1x score)
}

/**
 * Meld (exposed tile combination: pong/kong)
 *
 * Backend correspondence: @dataclass(frozen=True) Meld
 */
export interface Meld {
  meldType: MeldType;
  tiles: Tile[];        // Backend: Tuple[Tile, ...], frontend uses array
  isConcealed: boolean; // Whether it's concealed kong (used for men qing judgment)
}

/**
 * Utility function: Create Meld from backend data
 */
export function createMeld(data: {
  meld_type: string;
  tiles: Array<{ suit: string; rank: number }>;
  is_concealed?: boolean;
}): Meld {
  return {
    meldType: data.meld_type as MeldType,
    tiles: data.tiles.map(createTile),
    isConcealed: data.is_concealed || false
  };
}
